### PR TITLE
Go version of Plek

### PR DESCRIPTION
**Pull request by alext**
This introduces a [Go](http://golang.org) implementation of the Plek library.  I'd appreciate code review at this stage.

I'd also opinions on where this should live.  Should it live where it is now, in a separate repo, in the root of this repo?

### CODE REVIEW

**Comment by KushalP** [comment]
> I'd appreciate code review at this stage.

I think this looks good. That it has no external dependencies is also nice.

> I'd also opinions on where this should live. Should it live where it is now, in a separate repo, in the root of this repo?

Because this is going to be used as a library (and imported) by other services I'd like it to either live in this repo under a directory called `plek-go` or live in it's own repo at `alphagov/plek-go`. That way when another service calls it, the code would be `plek.New(...)`. Open to other ideas but putting it in a directory called `go` doesn't feel right.

**Comment by alext** [comment]
>Because this is going to be used as a library (and imported) by other services I'd like it to either live in this repo under a directory called `plek-go` or live in it's own repo at `alphagov/plek-go`. That way when another service calls it, the code would be `plek.New(...)`. Open to other ideas but putting it in a directory called go doesn't feel right.

I'd point out that the import path has no bearing on the default package name in consuming code.  That comes from the `package` declaration in the library.  This means that with an import path of `github.com/alphagov/plek/go`, you would still do `plek.New(...)`.

There are examples of other packages doing things like this; for example [vegeta](https://github.com/tsenart/vegeta) uses `github.com/tsenart/vegeta/lib` as it's import path for the library, but the package name is still `vegeta`.

**Comment by KushalP** [comment]
Oooh. I didn't know that. Then I'm happy for this to be merged.

File: go/plek.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return nil, &EnvVarURLInvalid{EnvVar: envVar, Err: err}
+	}
+	return u, nil
+}
+
+func serviceURLFromEnvOverride(serviceName string) (*url.URL, error) {
+	varName := "PLEK_SERVICE_" + strings.ToUpper(strings.Replace(serviceName, "-", "_", -1)) + "_URI"
```
Comment (mattbostock) [inline go/plek.go]: Could this use `fmt.Sprintf()` and be split into more variables, for neatness? It's quite hard to read as-is.

Comment (alext) [inline go/plek.go]: Fair point, would this be better:
```go
varName := fmt.Sprintf("PLEK_SERVICE_%s_URI", strings.ToUpper(strings.Replace(serviceName, "-", "_", -1)))
```
or do you think that's still too hard to read?  I'm not sure what I'd call the intermediate variable...

Comment (alext) [inline go/plek.go]: Actually, with that split over several lines it's much better - see d72849e.

Comment (mattbostock) [inline go/plek.go]: I think that's fine. I was thinking of something like:

```golang
serviceUpper := strings.ToUpper(strings.Replace(serviceName, "-", "_", -1)
varName := fmt.Sprintf("PLEK_SERVICE_%s_URI", serviceUpper)
```

Comment (mattbostock) [inline go/plek.go]: Agreed - d72849e is just right.

**Comment by alext** [comment]
It seems that there's a rough consensus that this is ok to live where it is.  I think this is therefore ready to be merged...

**Pull request merged**